Firestore/Source/Core/FSTQuery.mm

```diff
@@ -66,6 +68,43 @@ static constexpr NSComparisonResult ReverseOrder(NSComparisonResult result) {
   }
 }
 
+@implementation FSTFilter
+
++ (instancetype)createWithField:(const FieldPath &)field
+                             op:(FSTRelationFilterOperator)op
+                          value:(FSTFieldValue *)value {
+  if ([value isEqual:[FSTNullValue nullValue]]) {
+    if (op != FSTRelationFilterOperatorEqual) {
+      FSTThrowInvalidUsage(@"InvalidQueryException",
+                           @"Invalid Query. You can only perform equality comparisons on nil / "
+                            "NSNull.");
+    }
+    return [[FSTNullFilter alloc] initWithField:field];
+  } else if ([value isEqual:[FSTDoubleValue nanValue]]) {
+    if (op != FSTRelationFilterOperatorEqual) {
+      FSTThrowInvalidUsage(@"InvalidQueryException",
+                           @"Invalid Query. You can only perform equality comparisons on NaN.");
+    }
+    return [[FSTNanFilter alloc] initWithField:field];
+  } else {
+    return [FSTRelationFilter filterWithField:field filterOperator:op value:value];
+  }
+}
+
+- (const FieldPath &)field {
+  @throw FSTAbstractMethodException();  // NOLINT
```

Comment:
The docs for FSTAbstractMethodException suggest that adding `// NOLINT` is necessary... but a quick run of the linter without it seemed to pass.

After this PR is submitted, I'll check that a bit more thoroughly, and if it turns out to not be necessary, I'll remove it everywhere.

Reply:
The Google-internal objective-c linter is the one that complains. We've never managed to integrate it into our Objective-C build on github.

Reply:
Ah, interesting. I'll leave it as is then.